From b7333394d1a0dc94a02f7fd722914dce343f10c6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?D=C4=81vis=20Mos=C4=81ns?= <davispuh@gmail.com>
Date: Mon, 4 Dec 2023 02:38:19 +0200
Subject: [PATCH] kernelbase: Add test case for LoadLibraryEx() when hFile is
 non-NULL

When hFile is non-NULL then LoadLibraryEx() should return NULL
with LastError set as ERROR_INVALID_PARAMETER
---
 dlls/kernelbase/tests/file.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/dlls/kernelbase/tests/file.c b/dlls/kernelbase/tests/file.c
index 18dff0a32ab..b19360c8998 100644
--- a/dlls/kernelbase/tests/file.c
+++ b/dlls/kernelbase/tests/file.c
@@ -46,6 +46,23 @@ static void test_ioring_caps(void)
     todo_wine ok(hr == S_OK, "got %#lx.\n", hr);
 }
 
+static void test_load_library_ex_flags(void)
+{
+    HMODULE hmod;
+    DWORD last_error;
+
+    SetLastError(0x33);
+
+    /* Test if LoadLibraryEx can load a dll */
+    hmod = LoadLibraryExW(L"kernelbase.dll", NULL /* hFile */, LOAD_LIBRARY_SEARCH_SYSTEM32 /* dwFlags */);
+    ok(hmod != NULL, "LoadLibraryEx returned %p with LastError: %#lx.\n", hmod, GetLastError());
+
+    /* When hFile is non-NULL LoadLibraryEx should return NULL and set LastError to ERROR_INVALID_PARAMETER */
+    hmod = LoadLibraryExW(L"kernelbase.dll", (HANDLE)0x000003A0 /* random hFile */, 0 /* dwFlags */);
+    last_error = GetLastError();
+    ok(hmod == NULL && last_error == ERROR_INVALID_PARAMETER, "LoadLibraryEx returned %p with LastError: %#lx.\n", hmod, last_error);
+}
+
 START_TEST(file)
 {
     HMODULE hmod;
@@ -54,4 +71,5 @@ START_TEST(file)
     pQueryIoRingCapabilities = (void *)GetProcAddress(hmod, "QueryIoRingCapabilities");
 
     test_ioring_caps();
+    test_load_library_ex_flags();
 }
-- 
GitLab

